File: lib/data/fluxmapreader.cc
#include "lib/core/globals.h"
#include "lib/config/config.h"
#include "lib/data/fluxmap.h"
#include "lib/data/fluxmapreader.h"
#include "lib/data/fluxpattern.h"
#include "lib/config/proto.h"
#include "protocol.h"
#include <numeric>
#include <math.h>
#include <strings.h>

FluxmapReader::FluxmapReader(const Fluxmap& fluxmap):
    _fluxmap(fluxmap),
    _bytes(fluxmap.ptr()),
    _size(fluxmap.bytes()),
    _config(globalConfig()->decoder())
{
    rewind();
}

void FluxmapReader::getNextEvent(int& event, unsigned& ticks)
{
    ticks = 0;

    while (!eof())
    {
        uint8_t b = _bytes[_pos.bytes++];
        ticks += b & 0x3f;
        if (!b || (b & (F_BIT_PULSE | F_BIT_INDEX)))
        {
            _pos.ticks += ticks;
            event = b & 0xc0;
            return;
        }
    }

    _pos.ticks += ticks;
    event = F_EOF;
}

void FluxmapReader::skipToEvent(int event)
{
    unsigned ticks;
    findEvent(event, ticks);
}

bool FluxmapReader::findEvent(int event, unsigned& ticks)
{
    ticks = 0;

    while (!eof())
    {
        unsigned thisTicks;
        int thisEvent;
        getNextEvent(thisEvent, thisTicks);
        ticks += thisTicks;

        if (thisEvent == F_EOF)
            return false;

        if ((event == thisEvent) || (event & thisEvent))
            return true;
    }

    return false;
}

unsigned FluxmapReader::readInterval(nanoseconds_t clock)
{
    unsigned thresholdTicks =
        (clock * _config.pulse_debounce_threshold()) / NS_PER_TICK;
    unsigned ticks = 0;

    while (ticks <= thresholdTicks)
    {
        unsigned thisTicks;
        if (!findEvent(F_BIT_PULSE, thisTicks))
            break;
        ticks += thisTicks;
    }
    return ticks;
}

void FluxmapReader::seek(nanoseconds_t ns)
{
    unsigned ticks = ns / NS_PER_TICK;
    if (ticks < _pos.ticks)
    {
        _pos.ticks = 0;
        _pos.bytes = 0;
    }

    while (!eof() && (_pos.ticks < ticks))
    {
        int e;
        unsigned t;
        getNextEvent(e, t);
    }
    _pos.zeroes = 0;
}

void FluxmapReader::seekToByte(unsigned b)
{
    if (b < _pos.bytes)
    {
        _pos.ticks = 0;
        _pos.bytes = 0;
    }

    while (!eof() && (_pos.bytes < b))
    {
        int e;
        unsigned t;
        getNextEvent(e, t);
    }
    _pos.zeroes = 0;
}

nanoseconds_t FluxmapReader::seekToPattern(const FluxMatcher& pattern)
{
    const FluxMatcher* unused;
    return seekToPattern(pattern, unused);
}

nanoseconds_t FluxmapReader::seekToPattern(
    const FluxMatcher& pattern, const FluxMatcher*& matching)
{
    unsigned intervalCount = pattern.intervals();
    std::vector<unsigned> candidates(intervalCount + 1);
    std::vector<Fluxmap::Position> positions(intervalCount + 1);

    for (unsigned i = 0; i <= intervalCount; i++)
    {
        positions[i] = tell();
        candidates[i] = 0;
    }

    while (!eof())
    {
        FluxMatch match;
        if (pattern.matches(&*candidates.end(), match))
        {
            seek(positions[intervalCount - match.intervals]);
            _pos.zeroes = match.zeroes;
            matching = match.matcher;
            nanoseconds_t detectedClock = match.clock * NS_PER_TICK;
            if (detectedClock > (_config.minimum_clock_us() * 1000))
                return match.clock * NS_PER_TICK;
        }

        for (unsigned i = 0; i < intervalCount; i++)
        {
            positions[i] = positions[i + 1];
            candidates[i] = candidates[i + 1];
        }
        findEvent(F_BIT_PULSE, candidates[intervalCount]);
        positions[intervalCount] = tell();
    }

    matching = NULL;
    return 0;
}

void FluxmapReader::seekToIndexMark()
{
    skipToEvent(F_BIT_INDEX);
    _pos.zeroes = 0;
}

/*
 * Tries to guess the clock by finding the smallest common interval.
 * Returns nanoseconds.
 */
FluxmapReader::ClockData FluxmapReader::guessClock(
    double noiseFloorFactor, double signalLevelFactor)
{
    ClockData data = {};

    while (!eof())
    {
        unsigned interval;
        findEvent(F_BIT_PULSE, interval);
        if (interval > 0xff)
            continue;
        data.buckets[interval]++;
    }

    uint32_t max =
        *std::max_element(std::begin(data.buckets), std::end(data.buckets));
    uint32_t min =
        *std::min_element(std::begin(data.buckets), std::end(data.buckets));
    data.noiseFloor = min + (max - min) * noiseFloorFactor;
    data.signalLevel = min + (max - min) * signalLevelFactor;

    /* Find a point solidly within the first pulse. */

    int pulseindex = 0;
    while (pulseindex < 256)
    {
        if (data.buckets[pulseindex] > data.signalLevel)
            break;
        pulseindex++;
    }
    if (pulseindex == -1)
        return data;

    /* Find the upper and lower bounds of the pulse. */

    int peaklo = pulseindex;
    while (peaklo > 0)
    {
        if (data.buckets[peaklo] < data.noiseFloor)
            break;
        peaklo--;
    }

    int peakhi = pulseindex;
    while (peakhi < 255)
    {
        if (data.buckets[peakhi] < data.noiseFloor)
            break;
        peakhi++;
    }

    /* Find the total accumulated size of the pulse. */

    uint32_t total_size = 0;
    for (int i = peaklo; i < peakhi; i++)
        total_size += data.buckets[i];

    /* Now find the median. */

    uint32_t count = 0;
    int median = peaklo;
    while (median < peakhi)
    {
        count += data.buckets[median];
        if (count > (total_size / 2))
            break;
        median++;
    }

    /*
     * Okay, the median should now be a good candidate for the (or a) clock.
     * How this maps onto the actual clock rate depends on the encoding.
     */

    data.peakStart = peaklo * NS_PER_TICK;
    data.peakEnd = peakhi * NS_PER_TICK;
    data.median = median * NS_PER_TICK;
    return data;
}

File: java/com/cowlark/fluxengine/data/FluxmapReader.java
package com.cowlark.fluxengine.data;

import static com.cowlark.fluxengine.external.FluxEngine.F_BIT_INDEX;
import static com.cowlark.fluxengine.external.FluxEngine.F_BIT_PULSE;
import static com.cowlark.fluxengine.external.FluxEngine.F_EOF;
import static com.cowlark.fluxengine.external.FluxEngine.NS_PER_TICK;

import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.decoders.DecoderProto;
import java.time.Duration;

/**
 * A cursor over a Fluxmap's raw bytes.
 */
public class FluxmapReader
{
    private final Fluxmap fluxmap;
    private final Bytes bytes;
    private final int size;
    private final DecoderProto decoder;
    private int posBytes;
    private int posTicks;
    private int posZeroes;

    public FluxmapReader(Fluxmap fluxmap, DecoderProto decoder)
    {
        this.fluxmap = fluxmap;
        bytes = fluxmap.rawBytes();
        size = fluxmap.bytes();
        this.decoder = decoder;
        rewind();
    }

    public void rewind()
    {
        posBytes = 0;
        posTicks = 0;
        posZeroes = 0;
    }

    public boolean eof()
    {
        return posBytes == size;
    }

    public FluxPosition tell()
    {
        return new FluxPosition(posBytes, posTicks, posZeroes);
    }

    public void seek(FluxPosition pos)
    {
        posBytes = pos.bytes();
        posTicks = pos.ticks();
        posZeroes = pos.zeroes();
    }

    public Duration getDuration()
    {
        return Duration.ofNanos((long) (fluxmap.ticks() * NS_PER_TICK));
    }

    public int getCurrentEvent()
    {
        if (eof())
            return F_EOF;
        return bytes.getByte(posBytes) & 0xc0;
    }

    public Event getNextEvent()
    {
        long ticks = 0;
        while (!eof())
        {
            int b = bytes.getByte(posBytes++) & 0xff;
            ticks += b & 0x3f;
            if (b == 0 || (b & (F_BIT_PULSE | F_BIT_INDEX)) != 0)
            {
                posTicks += (int) ticks;
                return new Event(b & 0xc0, ticks);
            }
        }
        posTicks += (int) ticks;
        return new Event(F_EOF, ticks);
    }

    public void skipToEvent(int event)
    {
        findEvent(event);
    }

    public EventResult findEvent(int event)
    {
        long ticks = 0;
        while (!eof())
        {
            Event e = getNextEvent();
            ticks += e.ticks();
            if (e.event() == F_EOF)
                return new EventResult(false, ticks);
            if (event == e.event() || (event & e.event()) != 0)
                return new EventResult(true, ticks);
        }
        return new EventResult(false, ticks);
    }

    public long readInterval(long clockTicks)
    {
        long thresholdTicks = (long) (clockTicks * decoder.getPulseDebounceThreshold());
        long ticks = 0;
        while (ticks <= thresholdTicks)
        {
            EventResult r = findEvent(F_BIT_PULSE);
            if (!r.found())
                break;
            ticks += r.ticks();
        }
        return ticks;
    }

    public void seek(long ticks)
    {
        if (ticks < posTicks)
        {
            posTicks = 0;
            posBytes = 0;
        }
        while (!eof() && posTicks < ticks)
            getNextEvent();
        posZeroes = 0;
    }

    public void seekToByte(int b)
    {
        if (b < posBytes)
        {
            posTicks = 0;
            posBytes = 0;
        }
        while (!eof() && posBytes < b)
            getNextEvent();
        posZeroes = 0;
    }

    public void seekToIndexMark()
    {
        skipToEvent(F_BIT_INDEX);
        posZeroes = 0;
    }

    public ClockData guessClock()
    {
        return guessClock(0.01, 0.05);
    }

    public ClockData guessClock(double noiseFloorFactor, double signalLevelFactor)
    {
        ClockData data = new ClockData();
        while (!eof())
        {
            long intervalTicks = findEvent(F_BIT_PULSE).ticks();
            if (intervalTicks > 0xff)
                continue;
            data.buckets[(int) intervalTicks]++;
        }

        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;
        for (int b : data.buckets)
        {
            max = Math.max(max, b);
            min = Math.min(min, b);
        }
        data.noiseFloor = (int) (min + (max - min) * noiseFloorFactor);
        data.signalLevel = (int) (min + (max - min) * signalLevelFactor);

        int pulseindexTicks = 0;
        while (pulseindexTicks < 256)
        {
            if (data.buckets[pulseindexTicks] > data.signalLevel)
                break;
            pulseindexTicks++;
        }
        if (pulseindexTicks == 256)
            return data;

        int peakloTicks = pulseindexTicks;
        while (peakloTicks > 0)
        {
            if (data.buckets[peakloTicks] < data.noiseFloor)
                break;
            peakloTicks--;
        }

        int peakhiTicks = pulseindexTicks;
        while (peakhiTicks < 255)
        {
            if (data.buckets[peakhiTicks] < data.noiseFloor)
                break;
            peakhiTicks++;
        }

        int totalSize = 0;
        for (int i = peakloTicks; i < peakhiTicks; i++)
            totalSize += data.buckets[i];

        int count = 0;
        int medianTicks = peakloTicks;
        while (medianTicks < peakhiTicks)
        {
            count += data.buckets[medianTicks];
            if (count > totalSize / 2)
                break;
            medianTicks++;
        }

        data.peakStartTicks = peakloTicks;
        data.peakEndTicks = peakhiTicks;
        data.medianTicks = medianTicks;
        return data;
    }

    public record Event(int event, long ticks)
    {
    }

    public record EventResult(boolean found, long ticks)
    {
    }

    public static class ClockData
    {
        public long medianTicks;
        public int noiseFloor;
        public int signalLevel;
        public long peakStartTicks;
        public long peakEndTicks;
        public int[] buckets = new int[256];
    }
}
